Further improve D2k spiceblooms

reaperrr opened the pull request:
This implements the last missing aspects of spiceblooms, namely only growing/'spawning' on spice-free sand (unless pre-placed in map editor) and displaying the animation of blowing spice into the air when spawned.

Additionally, spice particles now properly target cells free of spice when possible.

Closes https://github.com/OpenRA/OpenRA/issues/9866.

reaperrr committed:
Only allow spice blooms to grow on certain terrain types

reaperrr committed:
Add invisible spice bloom spawn

When real spiceblooms explode, they spawn an invisible
spicebloom.spawnpoint actor, which only 'grows' when the cell is of
GrowthTerrainType. When 'exploding', it then displays the animation of
blowing spice up into the air, and spawns a real spicebloom.

This replicates the original behavior where spice blooms would only grow
on sand cells that had been harvested and free of spice.

reaperrr commented:
I intentionally kept this out from my previous PR because this leaves the 'fix' area and goes more into 'feature' territory and might need some polish because it's a little hacky.

Additionally, `var cells` in `SpiceBloom` probably still returns `null` all the time, but I don't know how to fix that.

pchote commented:
Do you really even want that random walk?  That behaviour is used in the other mods because it nicely simulates random growth from a central tree... in this case you're throwing the resources through the air, so you only really care that it will land in an empty cell within some reasonable circular range.

reaperrr committed:
Improve spice bloom seeding mechanism

Improve seeding target cell determination.
Draw out spice seedable cells before loop.
Default to null for SpiceBloom.Weapon.
Pull spice seeding into its own method.
Skip it completely if no seed weapon is defined.

reaperrr commented:
Updated.

Mailaender commented:
Very nice. Great attention to detail. This scales far better than infinite growth. :+1:

pchote commented on OpenRA.Mods.D2k/Traits/SpiceBloom.cs:
```diff
@@ -74,10 +81,18 @@ public SpiceBloom(ActorInitializer init, SpiceBloomInfo info)
 
 		public void Tick(Actor self)
 		{
+			if (!self.World.Map.Contains(spawnCell))
```
It would be better to use `self.Location` instead of caching `self.World.Map.CellContaining(self.CenterPosition)`.  The latter will break if the actor moves for whatever reason.

reaperrr replied:
Done.

pchote commented on OpenRA.Mods.D2k/Traits/SpiceBloom.cs:
```diff
@@ -97,10 +112,10 @@ public void Killed(Actor self, AttackInfo e)
 			{
-				var cells = OpenRA.Traits.Util.RandomWalk(self.Location, self.World.SharedRandom);
-				var cell = cells.Take(info.Range).SkipWhile(p => resLayer.GetResource(p) == resType && resLayer.IsFull(p)).Cast<CPos?>().RandomOrDefault(self.World.SharedRandom);
+				var cells = self.World.Map.FindTilesInAnnulus(self.World.Map.CellContaining(self.CenterPosition), 1, info.Range);
```
Why change from `self.Location`?

reaperrr replied:
Probably just misread it as self.CenterPosition originally. Anyway, fixed.

pchote commented:
We'll want this for prep, right?

penev92 commented:
I guess so.

penev92 commented on OpenRA.Mods.D2k/Traits/SpiceBloom.cs:
```diff
@@ -34,6 +37,9 @@ public class SpiceBloomInfo : ITraitInfo, Requires<RenderSpritesInfo>
 		public readonly string ResourceType = "Spice";
 
+		[Desc("Spice blooms only grow on this terrain type.")]
+		public readonly string GrowthTerrainType = "Sand";
```
Since we have 1 actor that uses the default and 1 that removes the default, I'd rather we default to "" (or `null`, whichever is appropriate according to our standards) and put "Sand" on the actor that needs it.

reaperrr replied:
Done.

penev92 commented on mods/d2k/weapons.yaml:
```diff
@@ -820,3 +820,12 @@ BloomExplosion:
+		Speed: 1c0
+		Blockable: false
+		Image: null
+	Warhead@3Eff: CreateEffect
```
This isn't the 3rd warhead on this weapon though.

reaperrr replied:
Fixed.

penev92 commented:
Looks good ingame.
:+1:  with the two nits above.

reaperrr commented:
Updated.

pchote commented on OpenRA.Mods.D2k/Traits/SpiceBloom.cs:
```diff
@@ -34,6 +37,9 @@ public class SpiceBloomInfo : ITraitInfo, Requires<RenderSpritesInfo>
 		public readonly string ResourceType = "Spice";
 
+		[Desc("Spice blooms only grow on this terrain type.")]
+		public readonly string GrowthTerrainType = null;
```
Can we make this an array instead, for flexibility?

atlimit8 replied:
`HashSet<string>`

pchote commented on OpenRA.Mods.D2k/Traits/SpiceBloom.cs:
```diff
@@ -97,10 +109,10 @@ public void Killed(Actor self, AttackInfo e)
 			{
-				var cells = OpenRA.Traits.Util.RandomWalk(self.Location, self.World.SharedRandom);
-				var cell = cells.Take(info.Range).SkipWhile(p => resLayer.GetResource(p) == resType && resLayer.IsFull(p)).Cast<CPos?>().RandomOrDefault(self.World.SharedRandom);
+				var cells = self.World.Map.FindTilesInAnnulus(self.Location, 1, info.Range);
```
You can pull this and the `SkipWhile` outside the loop to save some perf.

pchote commented on OpenRA.Mods.D2k/Traits/SpiceBloom.cs:
```diff
@@ -34,6 +37,9 @@ public class SpiceBloomInfo : ITraitInfo, Requires<RenderSpritesInfo>
+
 		[Desc("The weapon to use for spice creation.")]
 		[WeaponReference]
 		public readonly string Weapon = "SpiceExplosion";
```
Can we default this to null so that you don't need to override it with an empty entry in the spawnpoint?

RoosterDragon commented on OpenRA.Mods.D2k/Traits/SpiceBloom.cs:
```diff
@@ -78,6 +84,12 @@ public SpiceBloom(ActorInitializer init, SpiceBloomInfo info)
+			if (!self.World.Map.Contains(self.Location))
+				return;
+
+			if (info.GrowthTerrainTypes.Any() && info.GrowthTerrainTypes.Contains(self.World.Map.GetTerrainInfo(self.Location).Type))
```
Checking for `Any` here is redundant - did you mean to check something else? If not you can remove it.

pchote replied:
I think this should be `if (any && !contains)`, right?
We want to exit early if the current terrain type isn't in the set, but only if the set isn't empty.

pchote commented on OpenRA.Mods.D2k/Traits/SpiceBloom.cs:
```diff
@@ -91,45 +103,50 @@ public void Tick(Actor self)
-				pieces = info.Pieces[0];
-
-			for (var i = 0; pieces > i; i++)
+			if (!string.IsNullOrEmpty(info.Weapon))
```
`if (isnulllorempty) return;` would be cleaner.

reaperrr replied:
I think you're forgetting the spawning that starts at line 152 and is unrelated to the weapon.

pchote replied:
Ah, I missed that.  Still, this is a sign that it may want to be pulled out into a SeedResources method?

RoosterDragon commented on OpenRA.Mods.D2k/Traits/SpiceBloom.cs:
```diff
@@ -78,6 +84,12 @@ public SpiceBloom(ActorInitializer init, SpiceBloomInfo info)
+			if (!self.World.Map.Contains(self.Location))
+				return;
+
+			if (info.GrowthTerrainTypes.Any() && !info.GrowthTerrainTypes.Contains(self.World.Map.GetTerrainInfo(self.Location).Type))
```
Minor perf nit: Prefer `.Count > 0` since `Any` needs to allocate and this is called every tick.

pchote commented on OpenRA.Mods.D2k/Traits/SpiceBloom.cs:
```diff
@@ -91,45 +103,50 @@ public void Tick(Actor self)
 
-				var args = new ProjectileArgs
+				var cells = self.World.Map.FindTilesInAnnulus(self.Location, 1, info.Range)
+					.SkipWhile(p => resLayer.GetResource(p) == resType && resLayer.IsFull(p));
```
This may need to be `ToList`ed in order to get the benefits of it.

reaperrr replied:
Basically `.Where(p => resLayer.GetResource(p) != resType || !resLayer.IsFull(p)).ToList()`?

GraionDilach replied:
Also see #10062.

RoosterDragon replied:
`self.World.Map.FindTilesInAnnulus` will only find cells in the world by default, so you don't need the same check as #10062.

GraionDilach replied:
Ah, sorry then.

pchote commented on OpenRA.Mods.D2k/Traits/SpiceBloom.cs:
```diff
@@ -91,45 +103,50 @@ public void Tick(Actor self)
+				var cells = self.World.Map.FindTilesInAnnulus(self.Location, 1, info.Range)
+					.SkipWhile(p => resLayer.GetResource(p) == resType && resLayer.IsFull(p));
+
+				for (var i = 0; pieces > i; i++)
```
Style nit: `i < pieces`

RoosterDragon commented on OpenRA.Mods.D2k/Traits/SpiceBloom.cs:
```diff
@@ -91,45 +103,50 @@ public void Tick(Actor self)
-					Facing = 0,
+					var cell = cells.Cast<CPos?>().RandomOrDefault(self.World.SharedRandom);
+					if (cell == null)
+						cell = cells.Random(self.World.SharedRandom);
```
What are these three lines trying to do exactly? `RandomOfDefault` will return null only on an empty collection, so repeating it with `Random` will just throw which seems odd.

As written, you could simplify the current logic down to `cell = cells.Random(self.World.SharedRandom)` for the same effect.

reaperrr replied:
Tbh I didn't (and still don't) fully understand this piece of code from #8987, so I just avoided touching it too much.

RoosterDragon replied:
Looks like the old code was attempting to take a random cell in range but skipping over anything already full of resource. If there were no such cells (everything was already maxed) it just used any random cell in range instead.

Try moving the `SkipWhile` currently on the `cells` variable back onto the first random search line - that ought to make it work like the original code again as best I can tell.

obrakmann commented:
Looks fine to me. :+1:

Can you please squash those last four commits into the earlier ones before merging? Thanks!

reaperrr commented:
Squashed and reordered commits.

Note: There were some unavoidable yaml rebase conflicts while squashing. I'm quite sure i caught them all, but I still recommend giving this a last quick look-over before merging.

obrakmann commented:
Nope, still looks all good, in-game too.

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/958c08cff8820a3c01d0c3d0d26cf7b1fa4ab83a)
